File: Sources/Views/SplitFlap/SplitFlapModule.swift
import SwiftUI

/// A single split-flap character module. Animates through its character set
/// sequentially to reach the target — never jumps directly.
///
/// The animation replicates a physical split-flap mechanism:
/// 1. Top half of current character folds down (easeIn)
/// 2. Bottom half of next character swings into place (easeOut)
/// 3. Repeat until target character is reached
struct SplitFlapModule: View {
    let targetCharacter: Character
    let characterSet: [Character]
    var flipDuration: TimeInterval = 0.08
    var startDelay: TimeInterval = 0
    var theme: SplitFlapTheme = .classicAmber
    var animationSpeed: Double = 1.0
    var soundEnabled: Bool = true

    @State private var currentCharacter: Character = " "
    @State private var nextCharacter: Character = " "
    @State private var topAngle: Double = 0
    @State private var bottomAngle: Double = 90
    @State private var flipQueue: [Character] = []
    @State private var generation: Int = 0

    @Environment(\.accessibilityReduceMotion) private var reduceMotion

    private var effectiveDuration: TimeInterval {
        flipDuration / max(animationSpeed, 0.1)
    }

    var body: some View {
        GeometryReader { geo in
            let w = geo.size.width
            let h = geo.size.height
            let halfH = h / 2
            let fontSize = h * 0.65
            let inset: CGFloat = 3

            ZStack {
                // Module housing
                RoundedRectangle(cornerRadius: 4)
                    .fill(theme.flapBackground)
                    .shadow(color: .black.opacity(0.3), radius: 2, y: 1)

                VStack(spacing: 0) {
                    // === TOP HALF ===
                    ZStack {
                        // Static: top of next char (revealed as current top folds away)
                        characterHalf(nextCharacter, fontSize: fontSize,
                                      width: w - inset * 2, height: h - inset * 2, isTop: true)

                        // Animated: top of current char (folds down around hinge)
                        characterHalf(currentCharacter, fontSize: fontSize,
                                      width: w - inset * 2, height: h - inset * 2, isTop: true)
                            .rotation3DEffect(
                                .degrees(topAngle),
                                axis: (1, 0, 0),
                                anchor: .bottom,
                                perspective: 0.3
                            )
                            .opacity(topAngle < -89 ? 0 : 1)
                    }
                    .frame(height: halfH - 0.75)
                    .clipped()

                    // === HINGE ===
                    Rectangle()
                        .fill(theme.hingeColor)
                        .frame(height: 1.5)
                        .shadow(color: .black.opacity(0.4), radius: 0.5, y: 0.5)

                    // === BOTTOM HALF ===
                    ZStack {
                        // Static: bottom of current char (visible until covered)
                        characterHalf(currentCharacter, fontSize: fontSize,
                                      width: w - inset * 2, height: h - inset * 2, isTop: false)

                        // Animated: bottom of next char (swings from behind into place)
                        characterHalf(nextCharacter, fontSize: fontSize,
                                      width: w - inset * 2, height: h - inset * 2, isTop: false)
                            .rotation3DEffect(
                                .degrees(bottomAngle),
                                axis: (1, 0, 0),
                                anchor: .top,
                                perspective: 0.3
                            )
                            .opacity(bottomAngle > 89 ? 0 : 1)
                    }
                    .frame(height: halfH - 0.75)
                    .clipped()
                }
                .padding(inset)

                // Inner border for depth
                RoundedRectangle(cornerRadius: 4)
                    .strokeBorder(theme.hingeColor.opacity(0.3), lineWidth: 0.5)
            }
        }
        .aspectRatio(0.65, contentMode: .fit)
        .accessibilityLabel(Text(String(currentCharacter)))
        .onAppear {
            if reduceMotion {
                currentCharacter = targetCharacter
                nextCharacter = targetCharacter
            } else {
                scheduleFlipSequence()
            }
        }
        .onChange(of: targetCharacter) { _, _ in
            scheduleFlipSequence()
        }
    }

    // MARK: - Character Half Rendering

    /// Renders the top or bottom half of a character on a flap background.
    /// Uses frame alignment + clipping to split the character at the midpoint.
    private func characterHalf(_ char: Character, fontSize: CGFloat,
                               width: CGFloat, height: CGFloat, isTop: Bool) -> some View {
        Text(String(char))
            .font(.system(size: fontSize, weight: .bold, design: .monospaced))
            .foregroundColor(theme.characterColor)
            .frame(width: width, height: height)
            .background(theme.flapBackground)
            .frame(height: height / 2, alignment: isTop ? .top : .bottom)
            .clipped()
    }

    // MARK: - Animation Logic

    private func scheduleFlipSequence() {
        generation += 1
        let gen = generation

        if reduceMotion {
            currentCharacter = targetCharacter
            nextCharacter = targetCharacter
            return
        }

        // Snap to clean state before starting new sequence
        topAngle = 0
        bottomAngle = 90

        let queue = buildFlipQueue(from: currentCharacter, to: targetCharacter)
        guard !queue.isEmpty else { return }

        flipQueue = queue

        let delay = startDelay / max(animationSpeed, 0.1)
        Task { @MainActor in
            try? await Task.sleep(for: .milliseconds(Int(delay * 1000)))
            guard generation == gen else { return }
            processNextFlip(generation: gen)
        }
    }

    private func processNextFlip(generation gen: Int) {
        guard generation == gen, !flipQueue.isEmpty else { return }

        let next = flipQueue.removeFirst()
        nextCharacter = next
        topAngle = 0
        bottomAngle = 90

        let halfDuration = effectiveDuration * 0.5

        // Phase 1: Top half folds down
        withAnimation(.easeIn(duration: halfDuration)) {
            topAngle = -90
        } completion: {
            guard generation == gen else { return }

            // Phase 2: Bottom half swings into place
            withAnimation(.easeOut(duration: halfDuration)) {
                bottomAngle = 0
            } completion: {
                guard generation == gen else { return }

                // Play flap click when the flap lands
                if soundEnabled {
                    FlapSoundEngine.shared.playFlap()
                }

                // Advance to completed state
                currentCharacter = next
                nextCharacter = next
                topAngle = 0
                bottomAngle = 90

                // Continue with next character in queue
                processNextFlip(generation: gen)
            }
        }
    }

    /// Build the sequential queue of characters from current to target.
    /// Characters cycle forward only through the character set (like a real drum).
    func buildFlipQueue(from: Character, to: Character) -> [Character] {
        guard from != to else { return [] }
        guard let fromIdx = characterSet.firstIndex(of: from) else {
            return [to]
        }
        guard characterSet.contains(to) else {
            return [to]
        }

        var queue: [Character] = []
        var idx = fromIdx
        for _ in 0..<characterSet.count {
            idx = (idx + 1) % characterSet.count
            queue.append(characterSet[idx])
            if characterSet[idx] == to { break }
        }
        return queue
    }
}

// MARK: - Character Sets

extension SplitFlapModule {
    /// Digits 0-9, decimal point, and space (blank). Space first for drum wrap.
    static let digitCharacterSet: [Character] = {
        var chars: [Character] = [" "]
        chars.append(contentsOf: (0...9).map { Character(String($0)) })
        chars.append(".")
        return chars
    }()

    /// Uppercase A-Z and space. Space first for drum wrap.
    static let alphaCharacterSet: [Character] = {
        var chars: [Character] = [" "]
        chars.append(contentsOf: (65...90).map { Character(UnicodeScalar($0)) })
        return chars
    }()

    /// Full alphanumeric: space, 0-9, A-Z, decimal point.
    static let alphanumericCharacterSet: [Character] = {
        var chars: [Character] = [" "]
        chars.append(contentsOf: (0...9).map { Character(String($0)) })
        chars.append(contentsOf: (65...90).map { Character(UnicodeScalar($0)) })
        chars.append(".")
        return chars
    }()

    /// Suffix magnitudes: space, K, M, B, T.
    static let suffixCharacterSet: [Character] = [" ", "K", "M", "B", "T"]
}

File: Windows/src/components/SplitFlapModule.js
/**
 * SplitFlapModule — A single split-flap character module.
 * Port of SplitFlapModule.swift.
 *
 * Animates through its character set sequentially to reach the target —
 * NEVER jumps directly. The animation replicates a physical split-flap mechanism:
 * 1. Top half of current character folds down (easeIn)
 * 2. Bottom half of next character swings into place (easeOut)
 * 3. Repeat until target character is reached
 */
import { FlapSoundEngine } from '../audio/FlapSoundEngine.js';

export class SplitFlapModule {
  /**
   * @param {HTMLElement} container - Parent element to mount into
   * @param {string[]} characterSet - Ordered character set for this position
   * @param {object} [options]
   * @param {number} [options.flipDuration=0.08] - Total flip duration in seconds
   * @param {number} [options.startDelay=0] - Delay before starting a flip sequence
   * @param {number} [options.animationSpeed=1.0] - Speed multiplier
   * @param {boolean} [options.soundEnabled=true]
   */
  constructor(container, characterSet, options = {}) {
    this._container = container;
    this._characterSet = characterSet;
    this._flipDuration = options.flipDuration ?? 0.08;
    this._startDelay = options.startDelay ?? 0;
    this._animationSpeed = options.animationSpeed ?? 1.0;
    this._soundEnabled = options.soundEnabled ?? true;

    this._currentChar = ' ';
    this._nextChar = ' ';
    this._generation = 0;
    this._flipQueue = [];
    this._isFlipping = false;
    this._reducedMotion = window.matchMedia('(prefers-reduced-motion: reduce)').matches;

    window.matchMedia('(prefers-reduced-motion: reduce)').addEventListener('change', (e) => {
      this._reducedMotion = e.matches;
    });

    this._buildDOM();
    this._updateDisplay(this._currentChar, this._currentChar);
  }

  get effectiveDuration() {
    return this._flipDuration / Math.max(this._animationSpeed, 0.1);
  }

  /**
   * Set the target character. Triggers sequential flip animation.
   * @param {string} char
   */
  setTarget(char) {
    if (char === this._currentChar && !this._isFlipping) return;

    this._generation++;
    const gen = this._generation;

    if (this._reducedMotion) {
      this._currentChar = char;
      this._nextChar = char;
      this._updateDisplay(char, char);
      return;
    }

    const queue = this._buildFlipQueue(this._currentChar, char);
    if (queue.length === 0) return;

    this._flipQueue = queue;

    const delay = this._startDelay / Math.max(this._animationSpeed, 0.1);
    setTimeout(() => {
      if (this._generation !== gen) return;
      this._processNextFlip(gen);
    }, delay * 1000);
  }

  /** @param {boolean} enabled */
  setSoundEnabled(enabled) {
    this._soundEnabled = enabled;
  }

  /** @param {number} speed */
  setAnimationSpeed(speed) {
    this._animationSpeed = speed;
  }

  /** @param {boolean} reduced */
  setReducedMotion(reduced) {
    this._reducedMotion = reduced;
  }

  // ── DOM Construction ──

  _buildDOM() {
    const module = document.createElement('div');
    module.className = 'flap-module';

    const inner = document.createElement('div');
    inner.className = 'flap-module-inner';

    // Static top (next char, revealed as current folds away)
    this._staticTop = this._createHalf('top', 'static');
    // Animated top (current char, folds down)
    this._animTop = this._createHalf('top', 'animated');
    // Static bottom (current char, visible until covered)
    this._staticBottom = this._createHalf('bottom', 'static');
    // Animated bottom (next char, swings from behind)
    this._animBottom = this._createHalf('bottom', 'animated');

    // Hinge
    const hinge = document.createElement('div');
    hinge.className = 'flap-hinge';

    inner.appendChild(this._staticTop);
    inner.appendChild(this._animTop);
    inner.appendChild(this._staticBottom);
    inner.appendChild(this._animBottom);
    inner.appendChild(hinge);

    module.appendChild(inner);
    this._container.appendChild(module);
    this._module = module;
    this._inner = inner;
  }

  /**
   * @param {'top'|'bottom'} half
   * @param {'static'|'animated'} role
   * @returns {HTMLElement}
   */
  _createHalf(half, role) {
    const el = document.createElement('div');
    el.className = `flap-half flap-${half} ${role}`;

    const inner = document.createElement('div');
    inner.className = 'flap-half-inner';
    inner.textContent = ' ';

    // Set font size based on module height (will be set via CSS)
    el.appendChild(inner);
    return el;
  }

  /**
   * Update the displayed characters on all four flap halves.
   * @param {string} current
   * @param {string} next
   */
  _updateDisplay(current, next) {
    // Font size: 65% of module height
    const h = this._container.offsetHeight || 50;
    const fontSize = h * 0.65;

    const halves = [this._staticTop, this._animTop, this._staticBottom, this._animBottom];
    for (const half of halves) {
      half.querySelector('.flap-half-inner').style.fontSize = `${fontSize}px`;
    }

    // Static top shows NEXT char (revealed under folding current)
    this._staticTop.querySelector('.flap-half-inner').textContent = next;
    // Animated top shows CURRENT char (folds down)
    this._animTop.querySelector('.flap-half-inner').textContent = current;
    // Static bottom shows CURRENT char (visible until covered)
    this._staticBottom.querySelector('.flap-half-inner').textContent = current;
    // Animated bottom shows NEXT char (swings into place)
    this._animBottom.querySelector('.flap-half-inner').textContent = next;
  }

  // ── Animation Logic ──

  /**
   * Build the sequential flip queue from current to target character.
   * Cycles forward only through the character set (like a physical drum).
   * @param {string} from
   * @param {string} to
   * @returns {string[]}
   */
  _buildFlipQueue(from, to) {
    if (from === to) return [];

    const fromIdx = this._characterSet.indexOf(from);
    if (fromIdx === -1) return [to];
    if (!this._characterSet.includes(to)) return [to];

    const queue = [];
    let idx = fromIdx;
    for (let i = 0; i < this._characterSet.length; i++) {
      idx = (idx + 1) % this._characterSet.length;
      queue.push(this._characterSet[idx]);
      if (this._characterSet[idx] === to) break;
    }
    return queue;
  }

  /**
   * Process the next character in the flip queue.
   * Uses CSS transitions for the two-phase flip animation.
   * @param {number} gen - Generation counter to detect interruptions
   */
  _processNextFlip(gen) {
    if (this._generation !== gen || this._flipQueue.length === 0) {
      this._isFlipping = false;
      return;
    }

    this._isFlipping = true;
    const next = this._flipQueue.shift();
    this._nextChar = next;

    const halfDuration = this.effectiveDuration * 0.5 * 1000; // ms

    // Reset animated halves to starting positions
    this._animTop.style.transition = 'none';
    this._animTop.style.transform = 'rotateX(0deg)';
    this._animTop.style.opacity = '1';
    this._animBottom.style.transition = 'none';
    this._animBottom.style.transform = 'rotateX(90deg)';
    this._animBottom.style.opacity = '0';

    // Update display: current on animated top + static bottom, next on static top + animated bottom
    this._updateDisplay(this._currentChar, next);

    // Force reflow to apply reset
    void this._animTop.offsetHeight;

    // Phase 1: Top half folds down (easeIn)
    this._animTop.style.transition = `transform ${halfDuration}ms ease-in, opacity 1ms linear ${halfDuration - 1}ms`;
    this._animTop.style.transform = 'rotateX(-90deg)';
    this._animTop.style.opacity = '0';

    setTimeout(() => {
      if (this._generation !== gen) { this._isFlipping = false; return; }

      // Phase 2: Bottom half swings into place (easeOut)
      this._animBottom.style.opacity = '1';
      this._animBottom.style.transition = `transform ${halfDuration}ms ease-out`;
      this._animBottom.style.transform = 'rotateX(0deg)';

      setTimeout(() => {
        if (this._generation !== gen) { this._isFlipping = false; return; }

        // Play flap click when the flap lands
        if (this._soundEnabled) {
          FlapSoundEngine.shared.playFlap();
        }

        // Advance: next becomes current
        this._currentChar = next;
        this._updateDisplay(next, next);

        // Reset animated halves
        this._animTop.style.transition = 'none';
        this._animTop.style.transform = 'rotateX(0deg)';
        this._animTop.style.opacity = '1';
        this._animBottom.style.transition = 'none';
        this._animBottom.style.transform = 'rotateX(90deg)';
        this._animBottom.style.opacity = '0';

        // Continue with next character in queue
        this._processNextFlip(gen);
      }, halfDuration);
    }, halfDuration);
  }
}

// ── Character Sets ──

SplitFlapModule.DIGIT_SET = [' ', '0', '1', '2', '3', '4', '5', '6', '7', '8', '9', '.'];
SplitFlapModule.ALPHA_SET = [' ', ...Array.from({ length: 26 }, (_, i) => String.fromCharCode(65 + i))];
SplitFlapModule.SUFFIX_SET = [' ', 'K', 'M', 'B', 'T'];

/**
 * Per-position character sets for token display rows.
 * Positions 0-2: digits only
 * Position 3: space or decimal point
 * Positions 4-5: digits only
 * Position 6: digits + K/M/B/T suffixes
 */
SplitFlapModule.TOKEN_POSITION_SETS = (() => {
  const digitOnly = [' ', '0', '1', '2', '3', '4', '5', '6', '7', '8', '9'];
  const dotOnly = [' ', '.'];
  const suffixFlap = [...digitOnly, 'K', 'M', 'B', 'T'];
  return [
    digitOnly,  // pos 0
    digitOnly,  // pos 1
    digitOnly,  // pos 2
    dotOnly,    // pos 3
    digitOnly,  // pos 4
    digitOnly,  // pos 5
    suffixFlap, // pos 6
  ];
})();
